src/ucp/core/ucp_worker.c

```diff
@@ -638,8 +641,35 @@ ucp_worker_iface_error_handler(void *arg, uct_ep_h uct_ep, ucs_status_t status)
         for (lane = 0; lane < ucp_ep_num_lanes(ucp_ep); ++lane) {
             if ((uct_ep == ucp_ep->uct_eps[lane]) ||
                 ucp_wireup_ep_is_owner(ucp_ep->uct_eps[lane], uct_ep)) {
-                ret_status = ucp_worker_set_ep_failed(worker, ucp_ep, uct_ep,
-                                                      lane, status);
+                /* If the failure happened on AUX EP of CM lane on a server EP,
+                 * it means that client closed its CM_WIREUP_EP/AUX_EP and it
+                 * was detected before receiving WIREUP_MSG/ACK from a client or
+                 * marking a server's EP as REMOTE_CONNECTED was scheduled on a
+                 * progress, but not completed yet (CM_WIREUP_EP/AUX_EP is
+                 * closed when moving am EP to REMOTE_CONNECTED state) */
+                wireup_ep = ucp_wireup_ep(ucp_ep->uct_eps[lane]);
+                if ((lane == ucp_ep_get_cm_lane(ucp_ep))          &&
+                    (lane == ucp_ep_get_wireup_msg_lane(ucp_ep))  &&
+                    (wireup_ep != NULL)     &&
+                    ucp_wireup_aux_ep_is_owner(wireup_ep, uct_ep) &&
+                    (ucp_ep->flags & UCP_EP_FLAG_CONNECT_PRE_REQ_QUEUED)) {
+                    /* No need to invoke the error handling flow, just flush and
+                     * destroy CM_WIREUP/AUX_EP */
+                    aux_uct_ep = wireup_ep->aux_ep;
+                    ucs_assert(ucp_wireup_ep_test(aux_uct_ep));
+
+                    ucs_queue_head_init(&tmp_pending_queue);
+                    ucp_wireup_ep_disown(ucp_ep->uct_eps[lane], aux_uct_ep);
+                    ucp_worker_discard_uct_ep(ucp_ep->worker, aux_uct_ep,
+                                              UCT_FLUSH_FLAG_CANCEL,
+                                              ucp_wireup_pending_purge_cb,
+                                              &tmp_pending_queue);
+                    ucp_wireup_replay_pending_requests(ucp_ep, &tmp_pending_queue);
+                    ret_status = UCS_OK;
+                } else {
+                    ret_status = ucp_worker_set_ep_failed(worker, ucp_ep,
+                                                          uct_ep, lane, status);
+                }
```

Comment:
1. to helper function
2. can we remove some if conditions and replace by assertion?
3. do we really expect pending ops on aux ep?

Reply:
> yes, it has pending reqs scheduled when it was in `ucp_ep->uct_epslane[1]`

Which ops for example? Maybe we need to cancel them instead of replay?

Reply:
> Which ops for example? Maybe we need to cancel them instead of replay?

no, we couldn't cancel them
the requests are from user

Reply:
how could we have user requests on aux ep, I would expect they would be queued on wireup_ep and replayed on the final transport ep, and only after we complete the whole wireup process (including wireup req/reply)

Reply:
yes, correct
Those request are scheduled on a wireup EP, when it was created during CM phase.
So, when reconfiguration done on WIREUP_MSG phase

Reply:
> So, when reconfiguration done on WIREUP_MSG phase, those requests are on CM_WIREUP_EP/AUX_EP

Why are they not on the real transport? (the one which is created as a result of WIREUP_MSG)

Reply:
> Why are they not on the real transport? (the one which is created as a result of WIREUP_MSG)

We don't have a logic to move them from the AM lane of the previous configuration to the AM lane of the new configuration
After WIREUP_MSG is done, the users requests will be scheduled on the AM lane of the new configuration

Reply:
but why would any user message be added to AM lane of prev config?
isn't the AM lane of prev config a wireup_ep object, which does not allow sending on the real transport?

Reply:
> but why would any user message be added to AM lane of prev config?

since user did send an operation (e.g. `ucp_tag_send_nb()`), the operation was added to list of outstanding operations on the WIREUP_EP of the AM lane.
let's assume we had the following initial configuration:
```
lane[0]: tcp cm   [wireup_ep]
lane[1]: tcp/eth0 [wireup_ep] AM lane
```
and then after completing CM phase, it does WIREUP_MSG phase. Reconfiguration is done upon receiving WIREUP_MSG/REQ on a server and WIREUP_MSG/REPLY on a client.
let's assume server did reconfiguration and the new configuration is:
```
lane[0]: tcp cm   [wireup_ep: aux_ep="tcp/eth0"[wireup_ep]]
lane[1]: tcp/ib0  [wireup_ep]
lane[2]: tcp/ib1  [wireup_ep] AM lane
```
So, previously scheduled by user operations are on `CM_WIREUP_EP/AUX_EP` (`AUX_EP` is also WIREUP_EP for UCT/TCP EP).
The operation will be re-scheduled on AM lane of the new configuration, i.e. lane[2].

> isn't the AM lane of prev config a wireup_ep object, which does not allow sending on the real transport?

correct, it is a wireup object, and UCP will not send data (except PRE_REQ/REQ/REPLY WIREUP_MSGs)
the data will be sent after WIREUP_MSG phase completion on new AM lane

Reply:
thanks for the explanation, but something still does not feel right.. let's discuss it next week

Reply:
> let's discuss it next week

Sure, probably need to fix the logic and move the submitted requests to the new AM lane after reconfiguration

Reply:
@yosefe @brminich the PR has been updated to move the user's requests from the AUX_EP to AM lane of the new configuration and replay them after the reconfiguration is done